src/processor/document/candidate_relation_adder.py

```diff
@@ -0,0 +1,183 @@
+        added_relations_upper_bound_factor: int | None = None,
+        collect_statistics: bool = False,
+        sort_by_distance: bool = True,
+    ):
+        self.label = label
+        self.use_partition = use_partition
+        self.max_distance = max_distance
+        self.distance_type = distance_type
+        self.added_rels_upper_bound_factor = added_relations_upper_bound_factor
+        self.collect_statistics = collect_statistics
+        self.sort_by_distance = sort_by_distance
+        self.reset_statistics()
+
+    def __call__(
+        self,
+        document: DocumentWithEntitiesRelationsAndPartition,
+    ) -> DocumentWithEntitiesRelationsAndPartition:
+        available_relations = document.relations
+        available_relation_mapping = {(rel.head, rel.tail): rel for rel in available_relations}
+        if self.use_partition:
+            available_partitions = document.partition
+        else:
+            available_partitions = [Span(start=0, end=len(document.text))]
+        entities = document.entities
+        candidates_with_distance = {}
+        distances_taken = defaultdict(list)
+        num_relations_in_partition = 0
+        available_rels_within_allowed_distance = set()
+        for partition in available_partitions:
+            available_entities = [
+                entity
+                for entity in entities
+                if is_contained_in((entity.start, entity.end), (partition.start, partition.end))
+            ]
+            for head in available_entities:
+                for tail in available_entities:
+                    if head == tail:
+                        continue
+                    d = distance(
+                        (head.start, head.end),
+                        (tail.start, tail.end),
+                        self.distance_type,
+                    )
+                    if self.max_distance is not None and d > self.max_distance:
+                        continue
+                    if (head, tail) in available_relation_mapping:
+                        num_relations_in_partition += 1
+                        distances_taken[available_relation_mapping[(head, tail)].label].append(d)
+                        available_rels_within_allowed_distance.add(
+                            available_relation_mapping[(head, tail)]
+                        )
+                        continue
+                    candidates_with_distance[(head, tail)] = d
+        if self.sort_by_distance:
+            candidates_with_distance_list = sorted(
+                candidates_with_distance.items(), key=lambda item: item[1]
+            )
+        else:
+            candidates_with_distance_list = list(candidates_with_distance.items())
+            random.shuffle(candidates_with_distance_list)
+        n_added = 0
+        num_total_candidates = len(entities) * len(entities) - len(entities)
+        self.update_statistics("num_total_relation_candidates", num_total_candidates)
```

Comment:
can you check for `if self.collect_statistics` also inside of `update_statistics` methods?